api/src/main/java/org/apache/cloudstack/api/command/user/vm/CloneVMCmd.java

```diff
@@ -0,0 +1,193 @@
+    }
+
+    public void setTemporarySnapShotId(Long snapshotId) {
+        this.temporarySnapShotId = snapshotId;
+    }
+
+    public Long getTemporarySnapShotId() {
+        return temporarySnapShotId;
+    }
+
+
+    public void setTemporaryTemlateId(long tempId) {
+        this.temporaryTemlateId = tempId;
+    }
+
+    @Override
+    public void create() throws ResourceAllocationException {
+        try {
+            _userVmService.checkCloneCondition(this);
+            VirtualMachineTemplate template = _templateService.createPrivateTemplateRecord(this, _accountService.getAccount(getEntityOwnerId()), _volumeService);
+            if (template == null) {
+                throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, "failed to create a template to db");
+            }
+            s_logger.info("The template id recorded is: " + template.getId());
+            setTemporaryTemlateId(template.getId());
+            _templateService.createPrivateTemplate(this);
+            UserVm vmRecord = _userVmService.recordVirtualMachineToDB(this);
+            if (vmRecord == null) {
+                throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, "unable to record a new VM to db!");
+            }
+            setEntityId(vmRecord.getId());
+            setEntityUuid(vmRecord.getUuid());
+        }
+        catch (ResourceUnavailableException | InsufficientCapacityException e) {
+            s_logger.warn("Exception: ", e);
+            throw new ServerApiException(ApiErrorCode.RESOURCE_UNAVAILABLE_ERROR, e.getMessage());
+        } catch (InvalidParameterValueException e) {
+            s_logger.warn("Exception: ", e);
+            throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, e.getMessage());
+        } catch (ServerApiException e) {
+            throw new ServerApiException(e.getErrorCode(), e.getDescription());
+        } catch (CloudRuntimeException e) {
+            throw new ServerApiException(ApiErrorCode.INTERNAL_ERROR, e.getMessage());
+        } finally {
+            if (getTemporarySnapShotId() != null) {
+                _snapshotService.deleteSnapshot(getTemporarySnapShotId());
+                s_logger.warn("clearing the temporary snapshot: " + getTemporarySnapShotId());
+            }
+        }
+    }
+
+    public boolean isPublic() {
+        return false;
+    }
+
+    public String getVMName() {
+        return getTargetVM().getInstanceName();
+    }
+
+    public String getTemplateName() {
+        return (getVMName() + "-Clone-" + _uuidMgr.generateUuid(VirtualMachineTemplate.class, null)).substring(0, 32);
+    }
+
+    @Override
+    public void execute() {
+        Optional<UserVm> result;
+        try {
+            CallContext.current().setEventDetails("Vm Id for full clone: " + getEntityId());
+            s_logger.info("starting actual VM id: " + getEntityId());
+            result = _userVmService.cloneVirtualMachine(this, _volumeService, _snapshotService);
```

Comment:
check if volume and snapshot service objects are required to be passed here, or these can be accessed from the service layer?

Reply:
@atrocitytheme , I unresolved this comment as it is still valid. `VolumeService` is accessible and `SnapshotService` can easily be added to the `UserVmServiceImpl`. no reason to pass them.

Reply:
I see you did apply this in `UserVmManagerImpl` but didn't remove the parameters.